subsys/bluetooth/services/cts_c.c

```diff
@@ -0,0 +1,251 @@
+ */
+
+#include <sys/byteorder.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+
+#include <bluetooth/services/cts_c.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(cts_c, CONFIG_BT_GATT_CTS_C_LOG_LEVEL);
+
+/** The lowest valid Current Time year is the year when the Western calendar was introduced. */
+#define CTS_YEAR_MIN 1582
+/** The highest possible Current Time year. */
+#define CTS_YEAR_MAX 9999
+
+/* |     Year        |Month   |Day     |Hours   |Minutes |Seconds |Weekday |Fraction|Reason  |
+   |     2 bytes     |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  | = 10 bytes. */
+#define CTS_C_CURRENT_TIME_EXPECTED_LENGTH 10
+
+int bt_gatt_cts_c_init(struct bt_gatt_cts_c *cts_c,
+		       const struct bt_gatt_cts_c_init_param *cts_c_init)
+{
+	int err;
+
+	if (cts_c == NULL || cts_c_init == NULL) {
+		return -EINVAL;
+	}
+
+	memset(cts_c, 0, sizeof(struct bt_gatt_cts_c));
+
+	cts_c->evt_handler = cts_c_init->evt_handler;
+
+	err = k_sem_init(&cts_c->read_params_sem, 1, 1);
+
+	return err;
+}
+
+void bt_gatt_cts_c_on_disconnected(struct bt_gatt_cts_c *cts_c)
+{
+	k_sem_give(&cts_c->read_params_sem);
+}
+
+int bt_gatt_cts_c_handles_assign(struct bt_gatt_dm *dm,
+				 struct bt_gatt_cts_c *cts_c)
+{
+	const struct bt_gatt_dm_attr *gatt_service_attr =
+		bt_gatt_dm_service_get(dm);
+	const struct bt_gatt_service_val *gatt_service =
+		bt_gatt_dm_attr_service_val(gatt_service_attr);
+	const struct bt_gatt_dm_attr *gatt_chrc;
+	const struct bt_gatt_dm_attr *gatt_desc;
+
+	if (bt_uuid_cmp(gatt_service->uuid, BT_UUID_CTS)) {
+		return -ENOTSUP;
+	}
+	LOG_DBG("CTS found");
+
+	gatt_chrc = bt_gatt_dm_char_by_uuid(dm, BT_UUID_CTS_CURRENT_TIME);
+	if (gatt_chrc == NULL) {
+		return -EINVAL;
+	}
+	gatt_desc = bt_gatt_dm_desc_by_uuid(dm, gatt_chrc,
+					    BT_UUID_CTS_CURRENT_TIME);
+	if (gatt_desc == NULL) {
+		return -EINVAL;
+	}
+	cts_c->handle_ct = gatt_desc->handle;
+	gatt_desc = bt_gatt_dm_desc_by_uuid(dm, gatt_chrc, BT_UUID_GATT_CCC);
+	if (gatt_desc == NULL) {
+		return -EINVAL;
+	}
+	cts_c->handle_ct_ccc = gatt_desc->handle;
+	LOG_DBG("Current Time characteristic found.");
+
+	/* Finally - save connection object */
+	cts_c->conn = bt_gatt_dm_conn_get(dm);
+
+	return 0;
+}
+
+/**@brief Function for decoding a read from the Current Time characteristic.
+ *
+ * @param[in] p_time  Current Time structure.
+ * @param[in] p_data  Pointer to the buffer containing the Current Time.
+ * @param[in] length  Length of the buffer containing the Current Time.
+ *
+ * @retval 0        If the time struct is valid.
+ * @retval -EINVAL 	If the length does not match the expected size of the data.
+ */
+static int current_time_decode(current_time_char_t *p_time,
+			       uint8_t const *p_data, uint32_t const length)
+{
+	if (length != CTS_C_CURRENT_TIME_EXPECTED_LENGTH) {
+		/* Return to prevent accessing the out-of-bounds data. */
+		return -EINVAL;
+	}
+
+	uint32_t index = 0;
```

Comment:
Please move to the beginning of the function scope or you can simply increment pointer to data `p_data++` instead of creating index variable

Reply:
Done.